compiler/packages/babel-plugin-react-compiler/src/Optimization/LowerContextAccess.ts

```diff
@@ -0,0 +1,306 @@
+          return null;
+        }
+        keys.push(place.key.name);
+      }
+      return keys;
+    }
+  }
+}
+
+function emitPropertyLoad(
+  env: Environment,
+  obj: Place,
+  property: string,
+): {instructions: Array<Instruction>; element: Place} {
+  const loadObj: LoadLocal = {
+    kind: 'LoadLocal',
+    place: obj,
+    loc: GeneratedSource,
+  };
+  const object: Place = {
+    kind: 'Identifier',
+    identifier: makeTemporary(env.nextIdentifierId, GeneratedSource),
+    effect: Effect.Unknown,
+    reactive: false,
+    loc: GeneratedSource,
+  };
+  const loadLocalInstr: Instruction = {
+    lvalue: object,
+    value: loadObj,
+    id: makeInstructionId(0),
+    loc: GeneratedSource,
+  };
+
+  const loadProp: PropertyLoad = {
+    kind: 'PropertyLoad',
+    object,
+    property,
+    loc: GeneratedSource,
+  };
+  const element: Place = {
+    kind: 'Identifier',
+    identifier: makeTemporary(env.nextIdentifierId, GeneratedSource),
+    effect: Effect.Unknown,
+    reactive: false,
+    loc: GeneratedSource,
+  };
+  const loadPropInstr: Instruction = {
+    lvalue: element,
+    value: loadProp,
+    id: makeInstructionId(0),
+    loc: GeneratedSource,
+  };
+  return {
+    instructions: [loadLocalInstr, loadPropInstr],
+    element: element,
+  };
+}
+
+function emitSelectorFn(env: Environment, keys: Array<string>): Instruction {
+  const obj: Place = {
+    kind: 'Identifier',
+    identifier: {
+      id: makeIdentifierId(env.nextIdentifierId),
+      name: makeIdentifierName('c'),
+      mutableRange: {start: makeInstructionId(0), end: makeInstructionId(0)},
+      scope: null,
+      type: makeType(),
+      loc: GeneratedSource,
+    },
+    effect: Effect.Unknown,
+    reactive: false,
+    loc: GeneratedSource,
+  };
+  const instr: Array<Instruction> = [];
+  const elements = [];
+  for (const key of keys) {
+    const {instructions, element: prop} = emitPropertyLoad(env, obj, key);
+    instr.push(...instructions);
+    elements.push(prop);
+  }
+
+  const arrayInstr = emitArrayInstr(elements, env);
+  instr.push(arrayInstr);
+
+  const block: BasicBlock = {
+    kind: 'block',
+    id: makeBlockId(0),
+    instructions: instr,
+    terminal: {
+      id: makeInstructionId(0),
+      kind: 'return',
+      loc: GeneratedSource,
+      value: arrayInstr.lvalue,
+    },
+    preds: new Set(),
+    phis: new Set(),
+  };
+
+  const fn: HIRFunction = {
+    loc: GeneratedSource,
+    id: null,
+    fnType: 'Other',
+    env,
+    params: [obj],
+    returnType: null,
+    context: [],
+    effects: null,
+    body: {
+      entry: block.id,
+      blocks: new Map([[block.id, block]]),
+    },
+    generator: false,
+    async: false,
+    directives: [],
+  };
+
+  reversePostorderBlocks(fn.body);
+  removeUnreachableForUpdates(fn.body);
+  removeDeadDoWhileStatements(fn.body);
+  removeUnnecessaryTryCatch(fn.body);
+  markInstructionIds(fn.body);
+  mergeConsecutiveBlocks(fn);
+  enterSSA(fn);
+  inferTypes(fn);
```

Comment:
probably don't need some of these since it's a single block, is the plan to handle more complex cases?

Reply:
yeah i thought of keeping it in sync with the other passes that recreate blocks so if we add a new pass, we can just simply update this list 

delete unnecessary ones for now